src/Entity/SnapshotManager.php

```diff
@@ -231,60 +233,40 @@ public function cleanup(PageInterface $page, $keep)
         }
 
-        $tableName = $this->getTableName();
-        $platform = $this->getConnection()->getDatabasePlatform()->getName();
-
-        if ('mysql' === $platform) {
-            return $this->getConnection()->exec(sprintf(
-                'DELETE FROM %s
-                WHERE
-                    page_id = %d
-                    AND id NOT IN (
-                        SELECT id
-                        FROM (
-                            SELECT id, publication_date_end
-                            FROM %s
-                            WHERE
-                                page_id = %d
-                            ORDER BY
-                                publication_date_end IS NULL DESC,
-                                publication_date_end DESC
-                            LIMIT %d
-                        ) AS table_alias
-                )',
-                $tableName,
-                $page->getId(),
-                $tableName,
-                $page->getId(),
-                $keep
-            ));
+        $innerQb = $this->getRepository()->createQueryBuilder('i');
+        $expr = $innerQb->expr();
+
+        // try a better Function expression for this?
+        $ifNullExpr = sprintf(
+            'CASE WHEN %s THEN 1 ELSE 0 END',
+            $expr->isNull('i.publicationDateEnd')
+        );
+
+        // Subquery DQL doesn't support Limit
+        $innerQb
+            ->select('i.id')
+            ->where($expr->eq('i.page', $page->getId()))
+            ->orderBy($ifNullExpr, Criteria::DESC)
+            ->addOrderBy('i.publicationDateEnd', Criteria::DESC)
+            ->setMaxResults($keep);
+
+        $query = $innerQb->getQuery();
+        if (method_exists($query, 'getSingleColumnResult')) {
+            $innerArray = $query->getSingleColumnResult();
+        } else {
+            $innerArray = array_column($query->getArrayResult(), 'id');
         }
 
-        if ('oracle' === $platform) {
-            return $this->getConnection()->exec(sprintf(
-                'DELETE FROM %s
-                WHERE
-                    page_id = %d
-                    AND id NOT IN (
-                        SELECT id
-                        FROM (
-                            SELECT id, publication_date_end
-                            FROM %s
-                            WHERE
-                                page_id = %d
-                                AND rownum <= %d
-                            ORDER BY publication_date_end DESC
-                        ) table_alias
-                )',
-                $tableName,
-                $page->getId(),
-                $tableName,
-                $page->getId(),
-                $keep
+        $qb = $this->getRepository()->createQueryBuilder('s');
+        $expr = $qb->expr();
+        $qb->delete()
+            ->where($expr->eq('s.page', $page->getId()))
+            ->andWhere($expr->notIn(
+                's.id',
+                $innerArray
```

Comment:
Have you tried to use `$innerQb->getQuery()` here?

Reply:
yes i tried
> // Subquery DQL doesn't support Limit

Reply:
but i might try again inserting the final sql of the subquery as literal

Reply:
But it's fine insert the literal SQL, because it'll be generated by `doctrine`, the problem to use the sql manually is different sql syntax for different DB, and we need to add some `if/else` like before for `mysql` and `oracle`.

But in this case Doctrine will care about this and generate the correct sql, I guess it's better then go in the database twice!

Reply:
like i said before, i was using DQL and doctrine was okay with it, but that was ignoring Limit

i will try again with SQL, but i need to mark it so Doctrine isn't trying to parse it again

Reply:
But it's not a blocker for you PR @Hanmac, this command will be used rarely then it's fine we go in the database twice!

if you see the you will invest too much energy on that, please just ignore this :)

Reply:
doctrine doesn't seems to like to insert the generated SQL (with Limit) into the Query Builder for the outer Query.

a suggestion would be to add a DQL Function that gives the SubSelect the Limit back when turning DQL into SQL,
something like that:
https://stackoverflow.com/questions/24068947/subquery-with-limit-in-doctrine/46663949#46663949

Reply:
well I guess it's too much to be used only in one command, I guess if we need more subqueries in others places, we could use this approach :)